Fix markReplicasActive in DDLWorker and DatabaseReplicatedDDLWorker

tuanpach opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix markReplicasActive in DDLWorker and DatabaseReplicatedDDLWorker

When creating ephemeral nodes to indicate the replica is active, we hold the `ZooKeeperPtr` and `EphemeralNodeHolderPtr`. When the zookeeper expires, we expect the `DDLWorker::runMainThread` to throw an exception due to the expired keeper and re-mark the replica active in `markReplicaActive` (with `reinitialized=true`). However, the zookeeper may be recreated in other places. And `DDLWorker::runMainThread` does not encounter any error. So it doesn't remark the replica active. Hence, the active path in Keeper is removed, but the replica doesn't recreate the active path.
This PR checks if the holding keeper has expired to clear the corresponding `EphemeralNodeHolderPtr` in `markReplicaActive` even if  `reinitialized` is `false`. So that the active is recreated properly.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

tuanpach committed:
- When the keeper expires, remark the active_path for replica in  DDLWorker and DatabaseReplicatedDDLWorker.
- Reinitializing the main thread of DDLWorker if the zookeeper expired during event wait.
- Fix test_row_policy by not using 'localhost' in remote server configs to avoid conflicts when both nodes try to mark that host active.
- In DDLWorker::enqueueQueryAttempt and StorageSystemDDLWorkerQueue::fillData, only get zookeeper from context
- Only call getAndSetZookeeper in DDLWorker::initializeMainThread

tavplubix commented on src/Interpreters/DDLWorker.cpp:
```diff
@@ -1204,17 +1214,22 @@ void DDLWorker::runMainThread()
             subsequent_errors_count = 0;
 
             LOG_DEBUG(log, "Waiting for queue updates");
+            auto zookeeper = getAndSetZooKeeper();
             queue_updated_event->wait();
+            if (zookeeper->expired())
```
It shouldn't be expired right after calling `getAndSetZooKeeper` (it still can be, for example when the new session got expired immediately after creating it for some reason, but checking `zookeeper->expired()` right after `getAndSetZooKeeper` looks strange)

tuanpach commented:
> We still call `markReplicasActive` from only one place, `DDLWorker::runMainThread`:
> 
> https://github.com/ClickHouse/ClickHouse/blob/ee0849839709ba96b8823aa6593a680dfb15ac44/src/Interpreters/DDLWorker.cpp#L1197
> 
> Because of `active_node_holder_zookeeper->expired()` and `zk->expired()`, it will re-create `active` nodes on the next iteration of the main loop if someone called `getAndSetZooKeeper` from outside and updated the session without re-initializing DDLWorker. But it can take a while if the replica has a lot of entries to execute `scheduleTasks` (or if they contain heavy queries). But maybe it's not a big problem
> 
> > if someone called `getAndSetZooKeeper` from outside and updated the session without re-initializing DDLWorker
> 
> There are 2 such places: `DDLWorker::enqueueQueryAttempt` and `StorageSystemDDLWorkerQueue::fillData`. But maybe we don't need the exact same session there as in DDLWorker threads (looks like they work mostly independently from DDLWorker threads), so maybe we could just get a fresh session from `Context` in these places, while updating `DDLWorker::current_zookeeper` from the main thread only

I updated it, `DDLWorker::enqueueQueryAttempt` and `StorageSystemDDLWorkerQueue::fillData` now only get zookeeper from content.

tuanpach commented:
|||
|-|-|
|[00002_log_and_exception_messages_formatting](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=81395&sha=a1fe02fc4fc6a38b0f3653bbad429e19c702a761&name_0=PR&name_1=Stateless%20tests%20%28amd_asan%2C%20distributed%20plan%2C%20sequential%29)|https://github.com/ClickHouse/ClickHouse/issues/84642|
|[Logical error thrown (see clickhouse-server.log or logical_errors.txt)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=81395&sha=a1fe02fc4fc6a38b0f3653bbad429e19c702a761&name_0=PR&name_1=Stateless%20tests%20%28amd_binary%2C%20ParallelReplicas%2C%20s3%20storage%2C%20parallel%29)|https://github.com/ClickHouse/ClickHouse/issues/84771|
|[test_restore_replica/test.py::test_restore_replica_sequential](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=81395&sha=a1fe02fc4fc6a38b0f3653bbad429e19c702a761&name_0=PR&name_1=Integration%20tests%20%28asan%2C%20old%20analyzer%2C%205%2F6%29)|https://github.com/ClickHouse/ClickHouse/issues/84881|
|[Server died](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=81395&sha=a1fe02fc4fc6a38b0f3653bbad429e19c702a761&name_0=PR&name_1=Stress%20test%20%28amd_msan%29)|https://github.com/ClickHouse/ClickHouse/issues/84669|
